kubernetes-elasticsearch-cluster/create-cluster.sh

```diff
@@ -0,0 +1,64 @@
+#!/bin/bash
+
+set -o errexit
+set -o nounset
+
+help_text="Usage: kubernetes-elasticsearch-cluster/create-cluster.sh <dataset> <zone>
+  where <dataset> is the name of a directory in dataset_config/
+  <zone> is only needed if 'gcloud app create' hasn't yet been run in this project.
+Run this script from project root"
+
+if (( $# != 1 && $# != 2 )); then
+  echo "$help_text"
+  exit 1
+fi
+
+dataset=$1
+zone=""
+project_id=$(jq --raw-output '.project_id' dataset_config/${dataset}/deploy.json)
+echo "Creating Kubernetes cluster for dataset $dataset in project $project_id"
+
+gcloud config set project $project_id
+
+# Use the zone argument if present, otherwise try to lookup the App Engine location 
+# for this project.
+if (( $# == 2 )); then
+  zone=$2
+else
+  describe_text=$(gcloud app describe) || true
+  app_engine_region=$(echo "$describe_text" | grep locationId | awk '{print $2}')
+  # If no zone was found in App Engine, print the help text describing that it must
+  # be passed in as an argument to this script.
+  if [ -z "$app_engine_region" ]; then
+    echo "$help_text"
+    exit 1
+  fi
+
+  # Locations with only a single region (e.g. "us-central1") do not have the last 
+  # number returned in 'gcloud app describe' location (only "us-central"). 
+  # Check if the location is missing a digit as its last character and, if so,
+  # add a "1".
+  last_char=${app_engine_region: -1}
+  if [[ ! $last_char =~ ^[0-9]+$ ]]; then
+  	app_engine_region+="1"
+  fi
+  zone="${app_engine_region}-a"
+fi
+
+data_node_machine_type=$(jq --raw-output '.data_node_machine_type' dataset_config/${dataset}/deploy.json)
+if [ "$data_node_machine_type" == "null" ] || [ -z "$data_node_machine_type" ]; then
+	data_node_machine_type="n1-standard-1"
+fi
+
+data_node_replicas=$(jq --raw-output '.data_node_num_replicas' dataset_config/${dataset}/deploy.json) || "3"
+if [ "$data_node_replicas" == "null" ] || [ -z "$data_node_replicas" ]; then
+	data_node_replicas="3"
+fi
+
+# Use the cluster if it already exists, otherwise create it.
+gcloud container clusters describe elasticsearch-cluster >/dev/null || gcloud container clusters create elasticsearch-cluster --zone=${zone}
```

Comment:
Isn't this akin to your comment on creation of node pools where you didn't want the script to fail and instead delete/recreate (I don't think we should use the existing since `deploy.json` params may have changed).

I definitely would need to add a `--zone=${zone}` to the describe (I think I have a default zone set for gcloud which is why I've missed all these)

Reply:
This is more a question of for the next iteration (where we do not have separate node pools), should we delete and recreate the cluster, or fail if it already exists?

Reply:
Yeah, I was thinking of failing if it already exists. It's different from what I wanted with the other scripts. It's a bit arbitrary. I accidentally ran `create-cluster.sh` on the wrong dataset/project, and it was nice that it didn't delete the cluster. (I forgot about the new "Creating Kubernetes cluster..." line at the top, I'll look for that next time.)

Now you don't need separate `gcloud container node-pools` commands, you can just do:
```
gcloud container clusters create elasticsearch-cluster --zone=${zone} --machine-type XXX --num-nodes=XXX
```
So let's just do that (no describe), and it can fail if the cluster already exists.

Reply:
Done.